### ISSUE TITLE
A `\#` in the first argument of `\href` produces `/#` instead of `#`

### ISSUE THREAD

**Issue by pdmosses** [mathjax/MathJax#2565]
### Issue Summary

To define a macro `\mylink` with target id as parameter, such that `\mylink{a}` expands to `\href{#a}{a}`, the use of `#` in the macro definition needs to be escaped, to avoid ambiguity with LaTeX parameter references such as `#1` (and potentially `##1` for nested macros). 

Using `\#` works in a macro definition in LaTeX (when using the `hyperref` package):

```latex
\newcommand{\mylink}[1]{\href{\##1}{#1}}
```
but the same definition does not give the expected result in MathJax.

### Steps to Reproduce:

1. This is the first step

```markdown
$$
\newcommand{\mylink}[1]{\href{\##1}{#1}}
$$

Target math `\cssId{a}{A}`
$$
\cssId{a}{A}
$$

`\href{#a}{b}` gives a correct hyperlink to `#a`:

$$
\href{#a}{b}
$$

`\mylink{a}` gives an incorrect link to `localhost:4000/#a`:

$$
\mylink{a}
$$
```

2. This is the second step

Check that the link to the target from `b` is correct, and the link from `a` is broken.

I would expect MathJax to follow LaTeX by allowing the use of `\#` for `#` in the URL argument of `\href` in a macro definition. MathJax accepts the use of `\#`, but does not interpret it as `#` in the URL argument of `\href`.

### Technical details:

* MathJax Version: 3.1.2
* Client OS: macOS 10.15.7
* Browser: Safari 14.0
* Framework: Jekyll 4.1.1, using MathJax inside kramdown math blocks `$$...$$`

### Supporting information:

 * Include the MathJax configuration you are using, and the script tag that loads MathJax itself.

```
  <script>
    MathJax = { 
      tex: { 
        tags: 'ams',
        packages: {'[+]': ['textmacros']}
      },
      svg: { 
        fontCache: 'global', 
        displayAlign: 'left', 
        displayIndent: '2em',
        mtextInheritFont: true
      },
      loader: {
        load: ['[tex]/textmacros']
      }
    };
  </script>
  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-svg.js"></script>
```

I am a new user of MathJax. I am [experimenting](https://pdmosses.github.io/just-the-docs-tests/docs/js/mathjax.html) with using MathJax in the Jekyll theme [Just the Docs](https://github.com/pmarsceill/just-the-docs/discussions/432).

P.S. Here is the HTML generated by Jekyll/kramdown from the example in step 1 above:
```html
\[\newcommand{\mylink}[1]{\href{\##1}{#1}}\]
<p>Target math <code class="language-plaintext highlighter-rouge">\cssId{a}{A}</code>
\(\cssId{a}{A}\)</p>
<p><code class="language-plaintext highlighter-rouge">\href{#a}{b}</code> gives a correct hyperlink to <code class="language-plaintext highlighter-rouge">#a</code>:</p>
\[\href{#a}{b}\]
<p><code class="language-plaintext highlighter-rouge">\mylink{a}</code> gives an incorrect link to <code class="language-plaintext highlighter-rouge">localhost:4000/#a</code>:</p>
\[\mylink{a}\]
```

**Comment by pdmosses** [comment]
It appears that `\#` produces `/#` when used anywhere in the first argument of `\href`, regardless of whether that is in a macro definition. For example:
```html
\[\href{/docs/intro.html\#introduction}{introduction}\]
```
produces a link to `/docs/intro.html/#introduction`, as shown in the following screenshot:

<img width="407" alt="Screenshot" src="https://user-images.githubusercontent.com/18308236/98866971-1444fc80-246e-11eb-95eb-138cfe6c18ec.png">

So if that `/` produced by `\#` can be suppressed, `\href` should produce the correct link.

**Comment by pkra** [comment]
This is the current behavior: 

https://github.com/mathjax/MathJax-src/blob/fbd71f3f7ca06dc71304e8a2c03f13be233ccac5/ts/input/tex/html/HtmlMethods.ts#L42

I should probably try to process the string instead of just getting it.

**Comment by pdmosses** [comment]
I'm still puzzled about how `\#` produces a superfluous `/`. Perhaps there is a transformation that converts a `\` to a `/` in URLs (mapping Windows paths to Unix)?

**Comment by dpvc** [comment]
I've looked into the hyperref package in LaTeX, and it looks like it does process the URL (after manipulating the character catcodes and setting up some macros to handle `\#`, `\~` and some other characters).  We should probably try to duplicate that as closely as we can within MathJax's framework.

In the meantime, you can use

``` latex
\newcommand{\mylink}[1]{\href{###1}{#1}}
```

to get the result you are looking for.  (This won't work in actual LaTeX because of the catcodes, but should work in MathJax.)

> I'm still puzzled about how `\#` produces a superfluous `/`.  Perhaps there is a transformation that converts a `\` to a `/` in URLs (mapping Windows paths to Unix)?

If you use `\href{/docs/intro.html\#introduction}{introduction}`, the URL given to the browser is `/docs/intro.html\#introduction`, and it is the browser that does the transformation of the `\` to `/` to obtain `localhost:4000/docs/intro.html/#introduction`.

[Sorry for the long delay in replying.  I have been teaching my term on-line for the first time, and had zero extra time over the last several months.]

**Comment by pdmosses** [comment]
Many thanks for the clear diagnosis of this issue, and for the explanation of how to produce a unescaped `#` in a MathJax command definition.

I don't think it matters much that such a MathJax command definition won't work in LaTeX when using the hyperref package – especially as  the `\href` command differs also regarding its second argument (math vs text mode). The transformation from `##` to `#` actually looks quite natural in a URL argument, and is familiar from nested command definitions in (La)TeX. If a note is  added in the docs for the html extension, warning that `\href` doesn't remove `\`s in URLs, perhaps there would be no need to  duplicate the hyperref argument processing in MathJax?

**Comment by dpvc** [comment]
I've added the "Needs Documentation" label to remind me to either fix the URL processing or update the documentation.
